Core/include/Acts/Utilities/TransformRange.hpp

```diff
@@ -0,0 +1,133 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+namespace Acts {
+
+namespace detail {
+
+template <typename input_t>
+struct Dereference {
+  using value_type = decltype(*std::declval<input_t>());
+  constexpr static value_type apply(input_t& value) { return *value; }
+  constexpr static value_type apply(const input_t& value) { return *value; }
+};
+
+template <typename input_t>
+struct ConstDereference {
+  using value_type = decltype(std::as_const(*std::declval<input_t>()));
+
+  constexpr static value_type apply(input_t& value) { return *value; }
+  constexpr static value_type apply(const input_t& value) { return *value; }
+};
+
+template <template <typename> typename Callable, typename iterator_t,
+          bool force_const>
+struct TransformRangeIterator {
+ private:
+  using internal_value_type = typename iterator_t::value_type;
+
+ public:
+  using value_type = std::conditional_t<
+      force_const,
+      decltype(std::as_const(
+          std::declval<typename Callable<internal_value_type>::value_type>())),
```

Comment:
I would tend to use ```declval<typename Callable<internal_value_type>::value_type &>``` (this is at least what I saw used most and I think its otherwise an r-value or smth? Don't know how that works with ```as_const```)

Reply:
I'm forcing this to always be a reference in that case right? Is it a requirement on iterators that they return references when being dereferenced?

Reply:
I don't know of any such requirement. The enumerate view, I think, returns an r-value. Does ```as_const``` not also make it a reference (I am not sure what it does precisely)?

Reply:
I think `as_const` does you're right. I've gotten rid of that in this spot now.

Reply:
Also, now that I'm thinking about it, it is in fact a requirements, returning rvalues will result in dangling references now. 

I don't think I need to generalize this a this point to also support rvalues at this time, however.